File: J9Ch/src/S_4_Binary_Search_Sweep_Line/Required_10/_391_Number_of_Airplanes_in_the_Sky_Medium.py

"""
Definition of Interval.
class Interval(object):
    def __init__(self, start, end):
        self.start = start
        self.end = end
"""
def sorter(x, y):
    if x[0] != y[0]:
        return x[0] - y[0]
    return x[1] - y[1]

class Solution:
    # @param airplanes, a list of Interval
    # @return an integer
    def countOfAirplanes(self, airplanes):
        timepoints = []
        for airplane in airplanes:
            timepoints.append((airplane.start, 1))
            timepoints.append((airplane.end, -1))

        timepoints = sorted(timepoints, cmp=sorter)

        sum, most = 0, 0
        for t, delta in timepoints:
            sum += delta
            most = max(most, sum)

        return most

File: J9Ch/src/S_4_Binary_Search_Sweep_Line/Required_10/_391_Number_of_Airplanes_in_the_Sky_Medium.cpp

/**
 * Definition of Interval:
 * classs Interval {
 *     int start, end;
 *     Interval(int start, int end) {
 *         this->start = start;
 *         this->end = end;
 *     }
 */
class Solution {
public:
    /**
     * @param intervals: An interval array
     * @return: Count of airplanes are in the sky.
     */
    priority_queue<int,vector<int>,greater<int> >q;

    static bool comp(const Interval& a, const Interval& b){
        return a.start < b.start;
    }
    static int max(int a, int b) {
        return a > b ? a : b;
    }
    int countOfAirplanes(vector<Interval> &airplanes) {
        // write your code here
        sort(airplanes.begin(), airplanes.end(), comp);
        int result = 0;

        int len = airplanes.size();
        for (int i = 0; i < len; ++i) {
            while (q.size() > 0 && q.top() <= airplanes[i].start)
                q.pop();

            q.push(airplanes[i].end);
            result = max(result, q.size());
        }
        return result;
    }
};